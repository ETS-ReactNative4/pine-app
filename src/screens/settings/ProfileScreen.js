import React, { Component } from 'react';
import { ActionSheetIOS, Alert, StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { navigateWithReset, reset as resetApp } from '../../actions';
import * as settingsActions from '../../actions/settings';
import { handle as handleError } from '../../actions/error';
import { avatar } from '../../PinePaymentProtocol/user';
import getMnemonicByKey from '../../crypto/getMnemonicByKey';
import headerStyles from '../../styles/headerStyles';
import BackButton from '../../components/BackButton';
import SettingsTitle from '../../components/SettingsTitle';
import SettingsGroup from '../../components/SettingsGroup';
import SettingsLink from '../../components/SettingsLink';
import SettingsButton from '../../components/SettingsButton';
import StyledText from '../../components/StyledText';
import EditAvatar from '../../components/EditAvatar';
import CopyText from '../../components/CopyText';
import BaseSettingsScreen from './BaseSettingsScreen';

const styles = StyleSheet.create({
  profile: {
    paddingHorizontal: 15,
    marginBottom: 35,
    alignItems: 'center'
  },
  displayName: {
    fontSize: 26,
    textAlign: 'center',
    marginTop: 10,
    marginBottom: 3
  },
  address: {
    fontSize: 14,
    textAlign: 'center',
    color: '#8A8A8F'
  },
  signOutButtonContainer: {
    paddingRight: 0,
    marginLeft: 0
  },
  signOutButton: {
    alignSelf: 'center'
  },
  signOutButtonLoader: {
    right: null,
    alignSelf: 'center'
  }
});

@connect((state) => ({
  keys: state.keys.items,
  userProfile: state.settings.user.profile,
  hasCreatedBackup: state.settings.user.hasCreatedBackup,
  balance: state.bitcoin.wallet.balance
}))
export default class ProfileScreen extends Component {
  static navigationOptions = ({ navigation, screenProps }) => ({
    title: 'Profile',
    headerStyle: headerStyles.header,
    headerTitleStyle: headerStyles.title,
    headerLeft: (<BackButton onPress={() => { navigation.goBack(); }} />)
  });

  state = {
    signingOut: false
  };

  _flagAsUninitialized() {
    const dispatch = this.props.dispatch;

    const newSettings = {
      initialized: false,
      user: {
        hasCreatedBackup: false
      }
    };

    return dispatch(settingsActions.save(newSettings));
  }

  _getMnemonic() {
    const keys = Object.values(this.props.keys);
    const defaultKey = keys[0];

    return getMnemonicByKey(defaultKey.id);
  }

  _createManualBackup() {
    const navigation = this.props.navigation;

    return this._getMnemonic().then((mnemonic) => {
      navigation.navigate('BackUpMnemonic', {
        mnemonic,
        isModal: true
      });
    });
  }

  _canErase() {
    // Cannot remove a wallet with funds in it that hasn't been backed up.
    const { hasCreatedBackup, balance } = this.props;
    return balance === 0 || hasCreatedBackup;
  }

  _showWalletNotEmptyAlert() {
    Alert.alert(
      'Account Not Empty',
      'You cannot sign out from a non-empty account that has not been manually backed up. You would not have been able to sign back in again.',
      [
        { text: 'Create Manual Backup', onPress: this._createManualBackup.bind(this) },
        { text: 'Cancel', style: 'cancel' }
      ],
      { cancelable: false }
    );
  }

  _removeWallet(keepSettings) {
    const dispatch = this.props.dispatch;

    this.setState({ signingOut: true });

    return dispatch(resetApp(keepSettings))
      .then(() => {
        if (keepSettings) {
          this._flagAsUninitialized();
        }

        this.props.screenProps.dismiss();
        dispatch(navigateWithReset('Welcome'));
      })
      .catch((error) => {
        dispatch(handleError(error));
        this.setState({ signingOut: false });
      });
  }

  _showResetAppConfirmation() {
    if (!this._canErase()) {
      return this._showWalletNotEmptyAlert();
    }

    ActionSheetIOS.showActionSheetWithOptions({
      title: 'This will sign out your account from this device and iCloud. You can only sign back in if you have a manual backup of the recovery key.',
      options: ['Cancel', 'Sign Out', 'Create Manual Backup'],
      destructiveButtonIndex: 1,
      cancelButtonIndex: 0
    }, (buttonIndex) => {
      switch (buttonIndex) {
        case 1: // Erase Wallet and Settings.
          return this._removeWallet(false);

        case 2: // Create Manual Backup.
          return this._createManualBackup();
      }
    });
  }

  _showDisplayNameScreen() {
    const navigation = this.props.navigation;
    navigation.navigate('DisplayName');
  }

  _saveAvatarChecksum(checksum) {
    const { dispatch } = this.props;

    const newSettings = {
      user: {
        profile: {
          avatar: { checksum }
        }
      }
    };

    dispatch(settingsActions.save(newSettings));
  }

  _onSelectAvatar(image, error) {
    const { dispatch } = this.props;
    const { pineAddress } = this.props.userProfile;

    if (error) {
      return dispatch(handleError(error));
    }

    return this._getMnemonic().then((mnemonic) => {
      return avatar.set(pineAddress, image.data, mnemonic)
        .then((avatar) => {
          this._saveAvatarChecksum(avatar.checksum);
        })
        .catch((setError) => {
          dispatch(handleError(setError));
        });
    });
  }

  render() {
    const { userProfile } = this.props;
    const { pineAddress, displayName, avatar } = userProfile;
    const avatarChecksum = avatar ? avatar.checksum : null;

    return (
      <BaseSettingsScreen>
        <View style={styles.profile}>
          <EditAvatar
            onSelect={this._onSelectAvatar.bind(this)}
            pineAddress={pineAddress}
            checksum={avatarChecksum}
          />
          <View>
            <StyledText style={styles.displayName} numberOfLines={1}>{displayName}</StyledText>
            <CopyText copyText={pineAddress} underlayColor='#EFEFF3' tooltipArrowDirection='up'>
              <StyledText style={styles.address} numberOfLines={1}>{pineAddress}</StyledText>
            </CopyText>
          </View>
        </View>

        <SettingsTitle>Display Name</SettingsTitle>
        <SettingsGroup>
          <SettingsLink name={displayName} onPress={this._showDisplayNameScreen.bind(this)} isLastItem={true} />
        </SettingsGroup>

        <SettingsGroup>
          <SettingsButton
            title='Sign Out'
            type='destructive'
            onPress={this._showResetAppConfirmation.bind(this)}
            loading={this.state.signingOut}
            style={styles.signOutButton}
            containerStyle={styles.signOutButtonContainer}
            loaderStyle={styles.signOutButtonLoader}
            isLastItem={true}
          />
        </SettingsGroup>
      </BaseSettingsScreen>
    );
  }
}

ProfileScreen.propTypes = {
  keys: PropTypes.object,
  userProfile: PropTypes.object,
  hasCreatedBackup: PropTypes.bool,
  balance: PropTypes.number,
  screenProps: PropTypes.object,
  dispatch: PropTypes.func,
  navigation: PropTypes.any
};
