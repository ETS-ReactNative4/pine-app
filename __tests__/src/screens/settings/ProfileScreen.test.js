import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import ProfileScreen from '../../../../src/screens/settings/ProfileScreen';

jest.mock('../../../../src/containers/ErrorModalContainer', () => 'ErrorModalContainer');
jest.mock('../../../../src/components/BackButton', () => 'BackButtonBackButton');
jest.mock('../../../../src/components/SettingsTitle', () => 'SettingsTitle');
jest.mock('../../../../src/components/SettingsGroup', () => 'SettingsGroup');
jest.mock('../../../../src/components/SettingsLink', () => 'SettingsLink');
jest.mock('../../../../src/components/SettingsButton', () => 'SettingsButton');
jest.mock('../../../../src/components/StyledText', () => 'StyledText');
jest.mock('../../../../src/components/EditAvatar', () => 'EditAvatar');
jest.mock('../../../../src/components/CopyText', () => 'CopyText');

const storeMock = {
  getState: jest.fn(() => ({
    settings: {
      user: {
        hasCreatedBackup: false,
        profile: {
          avatar: {
            checksum: 'cb3e77ab-5cf3-4ff6-a599-e3eb45b2916d'
          }
        }
      }
    },
    bitcoin: {
      wallet: {
        balance: 0
      }
    },
    keys: {
      items: {
        'b62a06e9-7a1f-4862-a350-7af4505114c7': {
          id: 'b62a06e9-7a1f-4862-a350-7af4505114c7'
        }
      }
    }
  })),
  dispatch: jest.fn(),
  subscribe: jest.fn()
};

const navigationMock = {
  navigate: jest.fn(),
  state: {
    params: {}
  }
};

describe('ProfileScreen', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <ProfileScreen store={storeMock} navigation={navigationMock} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
