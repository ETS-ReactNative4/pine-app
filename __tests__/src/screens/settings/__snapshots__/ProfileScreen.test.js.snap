// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileScreen renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "flex": 1,
        "marginTop": 64,
      },
      Object {
        "backgroundColor": "#EFEFF3",
      },
    ]
  }
  theme={
    Object {
      "addContactIcon": Object {
        "testUri": "../../../src/images/icons/AddContact.png",
      },
      "backIcon": Object {
        "testUri": "../../../src/images/icons/Back.png",
      },
      "background": Object {
        "backgroundColor": "#ffffff",
      },
      "bigInput": Object {
        "color": "#000000",
      },
      "bigInputPlaceholderColor": "#C3C3C3",
      "bigInputSelectionColor": "#FFC431",
      "bubbleEndLeft": Object {
        "testUri": "../../../src/images/message/BubbleEndLeft.png",
      },
      "bubbleEndLeftError": Object {
        "testUri": "../../../src/images/message/BubbleEndLeftError.png",
      },
      "bubbleEndRight": Object {
        "testUri": "../../../src/images/message/BubbleEndRight.png",
      },
      "bubbleEndRightError": Object {
        "testUri": "../../../src/images/message/BubbleEndRightError.png",
      },
      "bubbleError": Object {
        "backgroundColor": "#FF3B30",
      },
      "bubbleErrorText": Object {
        "color": "#ffffff",
      },
      "bubbleReceived": Object {
        "backgroundColor": "#F0F0F0",
      },
      "bubbleReceivedText": Object {
        "color": "#000000",
      },
      "bubbleReceivedTextSmall": Object {
        "opacity": 0.75,
      },
      "bubbleSent": Object {
        "backgroundColor": "#FEC300",
      },
      "bubbleSentText": Object {
        "color": "#ffffff",
      },
      "bubbleSentTextSmall": Object {
        "opacity": 0.9,
      },
      "bulletPoint": Object {
        "backgroundColor": "#B1AFB7",
      },
      "button": Object {
        "backgroundColor": "#FFD23F",
      },
      "buttonDisabled": Object {
        "backgroundColor": "#B1AFB7",
      },
      "buttonLabel": Object {
        "color": "#ffffff",
      },
      "buttonLoaderColor": "#ffffff",
      "confirmTransactionDetail": Object {
        "borderColor": "#ECECEC",
      },
      "confirmTransactionErrorText": Object {
        "color": "#FF3B30",
      },
      "confirmTransactionHelpText": Object {
        "color": "#8E8E93",
      },
      "confirmTransactionLabel": Object {
        "color": "#8E8E93",
      },
      "confirmTransactionValue": Object {
        "color": "#000000",
      },
      "confirmTransactionView": Object {
        "backgroundColor": "#FAFAFA",
        "borderTopColor": "#F0F1F4",
      },
      "destructiveLabel": Object {
        "color": "#FF3B30",
      },
      "errorMessage": Object {
        "backgroundColor": "#FF3B30",
      },
      "errorText": Object {
        "color": "#FF3B30",
      },
      "feeLabelHighColor": "#FF3B30",
      "feeLabelMediumColor": "#FF9500",
      "header": Object {
        "backgroundColor": "#ffffff",
      },
      "headerButtonDisabled": Object {
        "color": "#E2E2E2",
      },
      "headerButtonPrimary": Object {
        "color": "#007AFF",
      },
      "headerSubtitle": Object {
        "color": "#B1AFB7",
      },
      "headerTitle": Object {
        "color": "#000000",
      },
      "homeGradientColors": Array [
        "rgba(255, 255, 255, 0.0)",
        "rgba(255, 255, 255, 1.0)",
      ],
      "illustrationBitcoinWithWings": Object {
        "testUri": "../../../src/images/illustrations/BitcoinWithWings.png",
      },
      "illustrationNoContacts": Object {
        "testUri": "../../../src/images/illustrations/EmptyListAvatar.png",
      },
      "illustrationNotification": Object {
        "testUri": "../../../src/images/illustrations/Notification.png",
      },
      "input": Object {
        "backgroundColor": "#F6F6F6",
        "borderColor": "#D8D8DB",
      },
      "inputBar": Object {
        "backgroundColor": "#F3F3F5",
      },
      "inputBarDisabledText": Object {
        "color": "#999999",
      },
      "inputBarErrorText": Object {
        "color": "#FF3B30",
      },
      "inputBarPlaceholderColor": "#999999",
      "inputBarSelectionColor": "#FFC431",
      "inputBarText": Object {
        "color": "#000000",
      },
      "inputCancelButton": Object {
        "backgroundColor": "#8A8A8F",
      },
      "inputDisabledButton": Object {
        "backgroundColor": "#D8D8DB",
      },
      "inputDisabledOverlay": Object {
        "backgroundColor": "#ffffff",
      },
      "inputFocus": Object {
        "borderColor": "#007AFF",
      },
      "inputSendButton": Object {
        "backgroundColor": "#FFD23F",
      },
      "label": Object {
        "color": "#B1AFB7",
      },
      "lightningBadgeImage": Object {
        "testUri": "../../../src/images/badges/LightningBadge.png",
      },
      "link": Object {
        "color": "#007AFF",
      },
      "mnemonicWord": Object {
        "color": "#999999",
      },
      "mnemonicWordInputErrorColor": "#FF3B30",
      "mnemonicWordInputSuccessColor": "#4CD964",
      "mnemonicWordNumber": Object {
        "color": "#999999",
      },
      "mnemonicWordWrapper": Object {
        "backgroundColor": "#F6F6F6",
        "borderColor": "#D8D8DB",
      },
      "name": "light",
      "offlineNoticeErrorColor": "#FF3B30",
      "offlineNoticeWarningColor": "#FF9500",
      "overlayCamera": Object {
        "backgroundColor": "#000000",
      },
      "overlayHome": Object {
        "backgroundColor": "#FEFEFE",
      },
      "overlayReceive": Object {
        "backgroundColor": "#FEFEFE",
        "borderColor": "gray",
      },
      "palette": Object {
        "background": "#ffffff",
        "background2": "#EFEFF3",
        "blue": "#007AFF",
        "blue2": "#3478F5",
        "foreground": "#000000",
        "gray1": "#B1AFB7",
        "gray10": "#C3C3C3",
        "gray11": "#E2E2E2",
        "gray12": "#A7A7AB",
        "gray13": "#C8C8CC",
        "gray14": "#C7C7CC",
        "gray15": "#CECED0",
        "gray2": "#8A8A8F",
        "gray3": "#BEBEBE",
        "gray4": "#D8D8DB",
        "gray5": "#999999",
        "gray6": "#F0F1F4",
        "gray7": "#ECECEC",
        "gray8": "#8E8E93",
        "gray9": "#9B9B9B",
        "green": "#4CD964",
        "indigo": "#5856d6",
        "middleground1": "#F6F6F6",
        "middleground2": "#F3F3F5",
        "middleground3": "#F0F0F0",
        "middleground4": "#FAFAFA",
        "middleground5": "#F2F2F6",
        "orange": "#FF9500",
        "orange2": "#FF8D36",
        "overlay": "#FEFEFE",
        "pink": "#ff2d55",
        "red": "#FF3B30",
        "yellow1": "#FFD23F",
        "yellow2": "#FFC431",
        "yellow3": "#FEC300",
        "yellow4": "#FFCD00",
      },
      "paragraph": Object {
        "color": "#B1AFB7",
      },
      "progressBarBackgroundColor": "#F6F6F6",
      "progressBarForegroundColor": "#FFCD00",
      "qrAvatar": Object {
        "borderColor": "#ffffff",
      },
      "qrBackground": "#ffffff",
      "qrForeground": "#000000",
      "qrWrapper": Object {
        "backgroundColor": "#ffffff",
        "shadowColor": "#000000",
      },
      "revealMnemonic": Object {
        "color": "#007AFF",
      },
      "sectionHeader": Object {
        "backgroundColor": "#F6F6F6",
      },
      "sectionHeaderText": Object {
        "color": "#8A8A8F",
      },
      "settingsArrow": Object {
        "color": "#C7C7CC",
      },
      "settingsBackground": Object {
        "backgroundColor": "#EFEFF3",
      },
      "settingsButtonDisabled": Object {
        "color": "#8A8A8F",
      },
      "settingsButtonPrimary": Object {
        "color": "#007AFF",
      },
      "settingsCheckmark": Object {
        "color": "#007AFF",
      },
      "settingsDescription": Object {
        "color": "#8A8A8F",
      },
      "settingsGreenCheckmark": Object {
        "color": "#4CD964",
      },
      "settingsGroup": Object {
        "backgroundColor": "#ffffff",
        "borderBottomColor": "#C8C8CC",
        "borderTopColor": "#C8C8CC",
      },
      "settingsHeader": Object {
        "backgroundColor": "#ffffff",
        "borderBottomColor": "#A7A7AB",
      },
      "settingsIcon": Object {
        "testUri": "../../../src/images/icons/Settings.png",
      },
      "settingsInput": Object {
        "color": "#000000",
      },
      "settingsItem": Object {
        "borderBottomColor": "#C8C8CC",
      },
      "settingsTitle": Object {
        "color": "#8A8A8F",
      },
      "settingsUnderlayColor": "#FAFAFA",
      "settingsValue": Object {
        "color": "#8E8E93",
      },
      "shareIcon": Object {
        "testUri": "../../../src/images/icons/Share.png",
      },
      "sliderTrackTintBackgroundColor": "#CECED0",
      "sliderTrackTintColor": "#3478F5",
      "smallLightningIconImage": Object {
        "testUri": "../../../src/images/icons/SmallLightning.png",
      },
      "stackedBarChart": Object {
        "backgroundColor": "#F2F2F6",
      },
      "stackedBarChartStack": Object {
        "borderLeftColor": "#ffffff",
      },
      "statusErrorColor": "#FF3B30",
      "statusSuccessColor": "#4CD964",
      "statusWarningColor": "#FF8D36",
      "subtitle": Object {
        "color": "#B1AFB7",
      },
      "subtitleUnread": Object {
        "color": "#000000",
      },
      "tableArrow": Object {
        "color": "#9B9B9B",
      },
      "tableBorder": Object {
        "borderColor": "#ECECEC",
      },
      "tableLabel": Object {
        "color": "#8E8E93",
      },
      "text": Object {
        "color": "#000000",
      },
      "title": Object {
        "color": "#000000",
      },
      "toolbarDot": Object {
        "backgroundColor": "#BEBEBE",
      },
      "toolbarMessagesIcon": Object {
        "testUri": "../../../src/images/icons/toolbar/Messages.png",
      },
      "toolbarReceiveIcon": Object {
        "testUri": "../../../src/images/icons/toolbar/Receive.png",
      },
      "toolbarSendIcon": Object {
        "testUri": "../../../src/images/icons/toolbar/Send.png",
      },
      "unitPicker": Object {
        "color": "#999999",
      },
      "walletBalanceInboundColor": "#FEC300",
      "walletBalanceOffChainColor": "#5856d6",
      "walletBalanceOnChainColor": "#4CD964",
      "walletBalancePendingColor": "#FF9500",
      "walletBalanceReservedColor": "#C7C7CC",
      "walletBalanceSpendableColor": "#4CD964",
      "walletBalanceUnredeemedColor": "#ff2d55",
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Array [
            Object {
              "paddingTop": 35,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "marginBottom": 35,
              "paddingHorizontal": 15,
            }
          }
        >
          <EditAvatar
            checksum="cb3e77ab-5cf3-4ff6-a599-e3eb45b2916d"
            onSelect={[Function]}
          />
          <View>
            <StyledText
              numberOfLines={1}
              style={
                Object {
                  "fontSize": 26,
                  "marginBottom": 3,
                  "marginTop": 10,
                  "textAlign": "center",
                }
              }
            />
            <CopyText
              tooltipArrowDirection="up"
              underlayColor="#EFEFF3"
            >
              <StyledText
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "fontSize": 14,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#8A8A8F",
                    },
                  ]
                }
              />
            </CopyText>
          </View>
        </View>
        <SettingsTitle>
          Display Name
        </SettingsTitle>
        <SettingsGroup>
          <SettingsLink
            isLastItem={true}
            onPress={[Function]}
          />
        </SettingsGroup>
        <SettingsTitle>
          Account Security
        </SettingsTitle>
        <SettingsGroup>
          <SettingsLink
            isLastItem={true}
            name="Recovery Key"
            onPress={[Function]}
          />
        </SettingsGroup>
        <SettingsGroup>
          <SettingsButton
            containerStyle={
              Object {
                "marginLeft": 0,
                "paddingRight": 0,
              }
            }
            isLastItem={true}
            loaderStyle={
              Object {
                "alignSelf": "center",
                "right": null,
              }
            }
            loading={false}
            onPress={[Function]}
            style={
              Object {
                "alignSelf": "center",
              }
            }
            title="Sign Out"
            type="destructive"
          />
        </SettingsGroup>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
